_posts/2024-01-26-cats-effect-parallelism.md

```diff
@@ -0,0 +1,789 @@
+
+```scala
+// other imports  
+import cats.syntax.all.*
+import cats.instances.all.*
+// other imports
+  
+object Main extends IOApp.Simple { 
+  
+  // other definitions
+
+  def routes(client: Client[IO], token: Token): HttpRoutes[IO] = {
+
+    given tk: Token = token
+
+    HttpRoutes.of[IO] {
+      case GET -> Root => Ok("hi :)")
+      case GET -> Root / "org" / orgName =>
+        for {
+          publicReposUri <- uri(publicRepos(orgName))
+          publicRepos <- fetch[PublicRepos](publicReposUri, client, PublicRepos.Empty)
+          pages = (1 to (publicRepos.value / 100) + 1).toVector
+          // let's calculate the vector of repository names
+          repositories <- pages.parUnorderedFlatTraverse { page =>
+            uri(repos(orgName, page)).flatMap(fetch[Vector[RepoName]](_, client, Vector.empty[RepoName]))
+          }
+          response <- Ok(repositories.toString)
+        } yield response
+    }
+  }
+}
+```
+
+Let me explain how we calculated `repositories` step by step:
+- The `parUnorderedFlatTraverse` function is used to perform parallel and unordered traversing of the `pages` vector. For each page, it fetches the repositories using the `uri(repos(orgName, page))` to construct the `URI` and `fetch[Vector[RepoName]](_, client, Vector.empty[RepoName])` to fetch the repository names from the GitHub API. The result is a `Vector[Vector[RepoName]]` representing repositories for each page.
+- The `parUnorderedFlatTraverse` combines the vectors from different pages into a single `Vector[RepoName]` by flattening and concatenating them in a parallel and unordered manner.
+- The resulting `Vector[RepoName]` is assigned to the repositories value.
+- Finally, the route responds with an HTTP Ok `Ok(repositories.toString)` containing the string representation of the vector of repository names.
+
+Let's manually test the output for different organizations:
+
+![alt ""](../images/github-contributors-aggregator/test_reponames.png)
+
+### 8.3 Start fetching contributors for each project in parallel and exhaust each iteratively
+
+Since we have all the repositories, now we can start fetching the contributors of each project in parallel. We must notice that the results for this one are also paginated, however, we don't know exactly how many contributors are out there for each project, it means that we're limited in parallelism here and we have to iteratively fetch "next contributors list" until it returns the less than 100 results, indicating that it's the last one.
+
+```scala
+// other imports  
+import cats.syntax.all.*
+import cats.instances.all.*
+// other imports
+  
+object Main extends IOApp.Simple { 
+  
+  // other definitions
+  def routes(client: Client[IO], token: Token): HttpRoutes[IO] = {
+    given tk: Token = token
+
+    HttpRoutes.of[IO] {
+
+      case GET -> Root => Ok("hi :)")
+
+      case GET -> Root / "org" / orgName =>
+        for {
+          publicReposUri <- uri(publicRepos(orgName))
```

Comment:
I think we should add some logs here - during the seconds we're waiting for a response, nothing seems to be happening

Reply:
Agreed, I will try adding logs for tracking the progress, from start to the end

Reply:
added informative logs, also warn and error logs where appropriate, please review.